Add exact and probabilistic modes section (#37873)
DataDog/documentation: hugo/content/en/observability_pipelines/processors/tag_cardinality_control.md

1. Define a filter query. See [Metrics Search Syntax][1] for more information.
    - Only metrics matching the filter are processed.
    - All metrics, regardless of whether they match the filter query, are sent to the next step in the pipeline.

> 1. Select a **Tracking mode** in the dropdown menu. See [Exact and probabilistic modes](#exact-and-probabilistic-modes) for more information.

Reviewer: ```suggestion
1. Select a **Tracking mode** in the dropdown menu. See the [Exact and probabilistic modes](#exact-and-probabilistic-modes) section for more information.
```

{{< img src="observability_pipelines/processors/tag_cardinality_control_overrides.png" alt="The per-metric override panel with a custom limit set to 100 with per-tag overrides for the host tag excluded from the limit and the region tag limited to five." style="width:80%;" >}}

## How the processor works

### Exact and probabilistic modes

The Tag Cardinality Control processor supports two modes for tracking tag cardinality:

- **exact**: Stores tag values as an 8-byte hash to optimize memory usage, at the cost of an extremely small chance of two distinct values being hashed to the same fingerprint.
- **probabilistic**: Uses [bloom filters][2] to track seen values, which can heavily optimize memory usage at the cost of occasional false positives. A false positive occurs when a value that has not been seen yet is incorrectly determined to have been seen, causing the processor to slightly exceed the specified cardinality limit.

> #### Memory usage for `exact` mode

Reviewer: exact and probabilistic are styled and capitalized in several different ways throughout the doc. would suggest removing the backticks in the header and bolding them in the text, and/or matching how they appear in the UI.

$$A = \text"total number of metrics"\ \×\ \text"average number of tag keys per metric"$$

$$B = \text"average length of each tag key"\ + (\text"value_limit"\ \×\ \text"average length of tag values")$$

$$\text"Memory usage" = A\ \×\ B$$

Since every tag value is stored as an 8-byte hash fingerprint, the `average length of tag values` is `8`.

#### Memory usage for probabilistic mode

Probabilistic mode uses bloom filters to track seen values for every (metric, tag) pair. For example, if the metric name is `request.latency` with tag keys `host` and `region`, the pairs tracked are:

- (`request.latency`, `host`)
- (`request.latency`, `region`)

The following formula estimates memory usage for probabilistic mode:

$$A = \text"total number of metrics"\ \×\ \text"average number of tag keys per metric"$$

$$B = \text"average length of field names for the tags"\ \+\ \text"cache_size_per_key"$$

$$\text"Memory usage" = A\ \×\ B$$

You can calculate `cache_size_per_key` with a Bloom Filter Calculator using a standard formula, where `n` is the cardinality limit and the false positive rate (`p`) is fixed at `0.1%` in the Worker.

> ### Benchmarks for exact mode vs probabilistic mode

Reviewer: ```suggestion
### Benchmarks for exact mode versus probabilistic mode
```

> The following tables show benchmarks for exact mode and probabilistic mode. As the number of unique values for each tag increases, probabilistic mode becomes more memory-efficient. The metric names and tag names used for these benchmarks were randomly generated 20-byte strings.

Author: ```suggestion
The following tables show benchmarks for exact mode and probabilistic mode. As the number of unique values for each tag increases, probabilistic mode becomes more memory efficient. The metric names and tag names used for these benchmarks were randomly generated 20-byte strings.
```

The machine type used for benchmarking is an AWS M6gd.4xlarge instance.

#### Benchmarks for 10,000 metrics tracked

> The Worker's RSS memory usage for probabilistic and exact mode for 10,000 metrics tracked.

Reviewer: ```suggestion
The following table shows the Worker's RSS memory usage for probabilistic and exact mode when tracking 10,000 metrics.
```

**10 tags per metric**

> | Unique values per tag | Probabilistic Mode (GB) | Exact Mode (GB) |

Reviewer: ```suggestion
| Unique values per tag | Probabilistic mode (GB) | Exact mode (GB) |
```

|:---------------------:|:-----------------------:|:---------------:|
| 500                   | 0.20                    | 1.04            |
| 250                   | 0.13                    | 0.56            |
| 100                   | 0.10                    | 0.21            |

#### Benchmarks for 50,000 metrics tracked

> The Worker's RSS memory usage for probabilistic and exact mode for 50,000 metrics tracked.

Reviewer: ```suggestion
The following table shows the Worker's RSS memory usage for probabilistic and exact mode when tracking 50,000 metrics.
```

**10 tags per metric**

> | Unique values per tag | Probabilistic Mode (GB) | Exact Mode (GB) |

Reviewer: ```suggestion
| Unique values per tag | Probabilistic mode (GB) | Exact mode (GB) |
```

|:---------------------:|:-----------------------:|:---------------:|
| 100                   | 0.35                    | 0.92            |
| 250                   | 0.54                    | 2.67            |
| 500                   | 0.83                    | 5.06            |

**50 tags per metric**

| Unique values per tag | Probabilistic Mode (GB) | Exact Mode (GB) |
|:---------------------:|:-----------------------:|:---------------:|
| 100                   | 1.27                    | 3.98            |
| 250                   | 2.08                    | 12.91           |
| 500                   | 3.33                    | 24.87           |

#### Benchmarks for 100,000 metrics tracked

> The Worker's RSS memory usage for probabilistic and exact mode for 100,000 metrics tracked.

Reviewer: ```suggestion
The following table shows the Worker's RSS memory usage for probabilistic and exact mode when tracking 100,000 metrics.
```

**10 tags per metric**

> | Unique values per tag | Probabilistic Mode (GB) | Exact Mode (GB) |

Reviewer: ```suggestion
| Unique values per tag | Probabilistic mode (GB) | Exact mode (GB) |
```

|:---------------------:|:-----------------------:|:---------------:|
| 100                   | 0.66                    | 1.79            |
| 250                   | 1.02                    | 5.29            |
| 500                   | 1.61                    | 10.11           |

**50 tags per metric**

> | Unique values per tag | Probabilistic Mode (GB) | Exact Mode (GB) |

Reviewer: ```suggestion
| Unique values per tag | Probabilistic mode (GB) | Exact mode (GB) |
```
